src/elexmodel/handlers/data/Featurizer.py

```diff
@@ -21,132 +23,147 @@ def __init__(self, features, fixed_effects):
+        return df
 
-    def featurize_heldout_data(self, heldout_data):
+    def filter_to_active_features(self, df: pd.DataFrame) -> pd.DataFrame:
         """
-        Featurize the data that the model will be applied on.
-        In our case the heldout_data is either the nonreporting_units
-        (when applying the model for the point predictions)
-        or conformalization_data/nonreporting_units
-        (when applying the model for the prediction intervals)
+        Get active features (ie. features + active fixed effects)
         """
-        new_heldout_data = heldout_data.copy()
+        return df[self.active_features]
 
-        if self.center_features:
-            self._center_features(new_heldout_data)
-
-        if self.add_intercept:
-            self._add_intercept(new_heldout_data)
-
-        if len(self.fixed_effect_cols) > 0:
-            missing_expanded_fixed_effects = []
-            new_heldout_data = self._expand_fixed_effects(new_heldout_data, drop_first=False)
-            # if all units from one fixed effect are reporting they will not appear in the heldout_data
-            # (e.g. nonreporting_units) and won't get a column when we expand the fixed effects
-            # on that dataframe. Therefore we add those columns with zero fixed effects manually.
-            # As an example, if we are running a county model using state fixed effects, and
-            # all of Delaware's counties are reporting, then no Delaware county will be in
-            # heldout_data (nonreporting_units), as a result there will be no column for Delaware
-            # in the expanded fixed effects of heldout_data (nonreporting_units).
-            for expanded_fixed_effect in self.expanded_fixed_effects:
-                if expanded_fixed_effect not in new_heldout_data.columns:
-                    missing_expanded_fixed_effects.append(expanded_fixed_effect)
-
-            missing_expanded_fixed_effects_df = pd.DataFrame(
-                np.zeros((new_heldout_data.shape[0], len(missing_expanded_fixed_effects))),
-                columns=missing_expanded_fixed_effects,
-            )
-            # if we use this method to add the missing expanded fixed effects because doing it manually
-            # ie. new_heldout_data[expanded_fixed_effect] = 0
-            # can throw a fragmentation warning when there are many missing fixed effects.
-            new_heldout_data = new_heldout_data.join(missing_expanded_fixed_effects_df)
-
-        return new_heldout_data[self.complete_features]
+    def generate_holdout_data(self, df: pd.DataFrame) -> pd.DataFrame:
+        """
+        Generate fixed effects for the holdout data (ie. data that we will predict on)
+        """
+        df = df.copy()
+
+        # if a unit has an inactive fixed effect value for some fixed effect category we need
+        # to insert 1 / (number of fixed effect values) into each active fixed effect value for that unit
```

Comment:
@lennybronner sorry, maybe an example will help.  Let's say this is my data:

| reporting_unit | reported | county_classification |
|---------------|---------:|----------------------|
| A                      | 1              | suburban                     |
| B                      | 1              | urban                           |
| C                      | 0             | rural                              |

We do `pd.get_dummies(drop_first=True)` on `county_classification` so let's say "suburban" gets dropped and we have:

| reporting_unit | reported | county_classification_urban | county_classification_rural |
|---------------|---------:|------------------------------|---------------------------|
| A                      | 1              | 0                                                | 0                                           |
| B                      | 1              | 1                                                 | 0                                          |
| C                      | 0             | 0                                                 | 1                                          |

Is the idea here that, for every `county_classification_*` that is 0, that `county_classification_*` will be 1/3 instead?  Or would that only be true for "C" since "A" and "B" are being used to train the model?

And if so, why 1/(number of county classifications)?  Isn't that saying "this county is 1/3rd urban" when in fact this county is actually not urban at all? 🤔 

Or am I way off-base and if so, please correct everything I've said lol 😅

Reply:
It would only be true for "C".
"A" is fine because it's suburban, so the intercept is effectively the dummy for that county.

re: why 1/n. We don't have a coefficient for rural counties, so we need to ask what our "best guess" for a rural county is. I think an average over the other county types is probably the best, unless you have a different suggestion? We can't leave it as zero (which is what it would be in the model where there are no rural counties in the training set) because that would be treating it like a suburban county. 

Does that make sense?

Reply:
Thanks, it does, but I still don't like it lol 😅   Especially the 1/n business.  In order to be more correct, I think we'd have to do some kind of fancy weighting/probability thing, or something... 😵‍💫

What I think we should do is, again if it's ok to use the entire data set to standardize features, it should be ok to use the entire data set to determine fixed effect columns.  Then, if you have two values, do `drop_first=True`, but more than two, `drop_first=False`.  Because now, let's say all of the "rural" counties wind up in the held-out set, as with our example: the `county_classification_rural` will be all 0's in the training set and thus dropped, right, which means it'll be dropped in the held-out set, too, which is fine.  `county_classification_suburban` and `county_classification_urban` will also be all 0's in the held-out set, which is fine since we're not fitting a model on the held-out set.

How does this sound?  I feel like I'm going to come back to this in an hour like "wait, no, that still won't work" 😅

Reply:
We do use the whole dataset to generate the fixed effects! That being said, I am not 100% sure I understand your suggestion 😬

Reply:
Oh!  I'm so sorry, I thought you were doing the training and held-out sets separately 🙈 

In that case, per fixed effect, I think doing `drop_first=True` when the number of fixed effect categories (_n_) is 2 and `drop_first=False` when _n_ > 2 makes sense.  Because then with the previous example, we'll have the training set:

| reporting_unit | reported | county_classification_urban | county_classification_suburban | intercept |
|---------------|---------:|-----------------------------:|--------------------------------:|----------:|
| A                      | 1              | 0                                                | 1                                                    | 1               |
| B                      | 1              | 1                                                 | 0                                                   | 1               |

`county_classification_rural` gets dropped with the addition of the `intercept` and on account of being all 0's, then the held-out set:

| reporting_unit | reported | county_classification_urban | county_classification_suburban | intercept |
|---------------|---------:|-----------------------------:|--------------------------------:|----------:|
| C                      | 0             | 0                                                | 0                                                   | 1                |

Honestly, I think that's all we can do.  Even if we have the situation where the training set is say 100% "suburban" so we drop both "rural" and "urban", I don't think we'll be able to do much better until more reporting units report. 😕 

Or, in regards to this particular fixed effect, maybe adding a feature for population density and/or registration will help, since I think that's at least part of what `county_classification` is trying to capture.  For instance, I suspect the `county_classification` for Arizona is either "Maricopa" and "not Maricopa" or "Maricopa", "rural", and "less rural" lol.  Either way, Maricopa takes forever to report, so we'll be going through tab for days with a model that's doing alright because of the baseline and demographics, but even then, Maricopa will have had a huge population growth between 2020 and 2024 and the model's missing out on that information (since maybe the demographic shift has been less dramatic) 🤔

Reply:
I'm still not quite sure I understand what differentiating between n=2 and n>2 does here? 

In your example, your design matrix is singular now. We'll need to drop either urban or suburban (and not rural, since that doesn't appear in the training data).

Reply:
Hmmmmmmmmm...good point, and the more I think about it, this could still happen even with `drop_first=True` 😭 

Again, if you have anything I can read about the 1/_n_ solution, that would be really helpful.  I'm unfamiliar with it and I'm not sure how it solves this problem 😞

Reply:
I tried looking for something to read on this, but I wasn't really able to find anything. In some way this is just a modeling choice (I had misunderstood Jen's question initially). If we have a categorical variable (and we are modeling with intercept, so we need to drop one of the categories) and one of the categories does not appear in the training set, what this means is that we know nothing about the category -- so what we do in that case is really our choice. I see two options:
1. We take the average category effect from the training data (which is what we are doing, and which is where the 1/n comes from)
2. We zero out everything (including the intercept). 

I would prefer solution (1) because I think removing the intercept will have other weird downstream effects for those rows. In some way it's a empirical question and we could eventually try them both and see which performs better.

Reply:
Thanks, yeah, there's a lot of stuff to think about here.  I agree that (2) could lead to weirdness, but I also agree that an evaluation would be helpful here; although I also suspect the weirdness brought on by (2) could be rare, so we'd have a hard time capturing it during evaluations 😞 

I'm also worried about a feature indicating that a county is "1/3rd suburban" and "1/3rd urban" when in fact it's entirely "rural" and "0 suburban", "0 urban".  Although, this is based on the data we currently have on hand, which is why I think adding in features for population density (or something similar) will help.  There are counties in New Jersey, for instance, that literally are ~1/3rd urban/suburban/rural (e.g. Passaic) and we're missing that with or without this change 😄 

A long-term thing to think about: I was doing some reading on this and discovered that apparently, [`scikit-learn`'s regression implementations use SVD instead of the matrix inversion](https://datascience.stackexchange.com/a/85870/77642).  So, rather than having an issue with multicollinearity, you just get weights that might be weird for explainability purposes, usually one of these features ending up with a weight of 0.  Unfortunately I'm not super familiar with the pros and cons of either approach, but I figured I'd share that anyway, if either switching to `scikit-learn` or re-writing `elex-solver` to use SVD was something we'd want to consider long-term 🤷🏻‍♀️

Reply:
Alright, so I'm going to mark this conversation as resolved and come back to these questions at a later date.  I've done my best to summarize this in ELEX-3252 which is now in the backlog (and currently assigned to me, but feel free to reassign as necessary).  But please add any additional notes/points of clarification/etc. to the ticket.  Thanks 😄